ATL-6 Create Event

michaelseyo opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Part 3 (Latest: 16/2/23):
- Create hamburger navbar at the top of create events page with additional sign out functionality

Part 2:
- Integrate backend with new db schema. For the creation of `event`, connects existing `Departments` and `User` data upon creation of an `event`.

Part 2 Notes:
- Created some data in the db for `Departments` so that I can reference upon creating an event.
- Used uuid for `eventId` 

Part 2 Future Improvements:
- In the schema e.g the`Departments` model has a `relation` with `Event` but it is currently 1-to-1. So in the event that we create a new `event` with departments that are already tied to an event, existing event tied to the department will be overriden and only the latest event is shown. Could improve to a 1-to-many in the future. This is also the same case for `User` as it is currently a 1-to-1 relationship with `Event`
- How do we create columns in railway that are of our custom type we have defined e.g Departments[] property for event

Part 1
- Create event frontend based off Evan's work, added form logic and validation
- Add create event backend that creates an event and saves to db, based off ziyuan's work

Part 1 Notes:
- Form validation for department checkboxes for scenario when submitting no selected departments still displays default error message by zod not very sure how to change it, must be due to it submitting an empty array: `[]` which is identified as `boolean` thus displaying message "Expected array, received boolean"
- Form logic and validation for `attendees` is abit hacky since I had to manually create it instead of integrating with `react-hook-forms`, as the `IndeterminateCheckbox` component for the `DataTable` requires an `onChange`  for its toggle functionality. This clashes with react-hook-forms' `register` function that implements its own `onChange`, such that I cannot register `attendees` into `react-hook-forms` 
- Typing is not very robust at the moment

Fixes # (issue)
ATL-6

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

woowenjun99 commented:
Hi Michael, thanks for the great work :). Here are some of the changes that in terms of the functionality aspect that I may require you to change:

- [ ] As of one of the PR and the changes proposed in the routing, I might require you to help me rebase from the main branch and fix the routing accordingly.

woowenjun99 commented on src/components/auth/RestrictedPage.tsx:
```diff
@@ -0,0 +1,22 @@
+const RestrictedPage = () => {
+  return (
+    <>
+      <Head>
```
This component looks good. Here are some to-dos for this page

- [ ] A page should be found under the `/page` directory. If you want, you can treat it as a page and put it under the corresponding directory, add a button redirecting the user back to the home page should they not have the required access level.
- [ ] Alternatively, you may treat it as a component. If this is so, do remove the `<Head></Head>` and rename the function correspondingly.

woowenjun99 commented on src/server/trpc/router/admin/events/create.ts:
```diff
@@ -0,0 +1,28 @@
+        },
+      })
+    } catch (error) {
+      console.log(error)
```
- [ ] Use `console.error((error as Error).message)` instead of `console.log()`, 
- [ ] If you want, you can throw a TRPCError with the corresponding code, description and message. Otherwise, remove the import from line 3.

woowenjun99 commented on src/pages/events/create.tsx:
```diff
@@ -0,0 +1,202 @@
+  const [attendees, setAttendees] = useState<string[]>([])
+  const [submitBefore, setSubmitBefore] = useState<boolean>(false) // hacky use for attendees validation
+  const { data: session } = useSession()
+  const { data: data } = trpc.event.getAllUsers.useQuery()
```
- [ ] No need `{data: data}`, destructuring `const {data} = useSession()` will suffice.
- [ ] Do add in a `useSession({ required: true })` as it will require the user to be logged in prior to accessing the page.

woowenjun99 commented on src/pages/events/create.tsx:
```diff
@@ -0,0 +1,202 @@
+                bgColor="#4365DD"
+                width={150}
+                type="submit"
+                disabled={isSubmitting}
```
- [ ] Is there any reason for the addition of the react-form dependency? The alternative would be `newEvent.isSubmitting`, `newEvent.error`, `newEvent.isSuccess` as of the TRPC Mutation Hook.

michaelseyo replied:
Didn't realise that there was one for submission my bad will use TRPC mutation instead

woowenjun99 reviewed (changes requested):
Hi Michael, I have tested the website and it works. In general, the design looks ok. 

**Coding POV (In-Depth)**
In terms of code, I have raised up some issues that I have went in depth into. 

- [x] Fix the issues raised

**Coding POV (High Level)** 

- [ ] Would it be possible to refactor the code into the separate components? I noticed that the two files are quite big. Do use the `/components` directory if possible to keep the `/page` neat.

- [x] Would it be possible to write some comments or documentation, especially in `DataTable.tsx` and `MultiSelectFilter.tsx` as I noticed that each function is very long and difficult to read. Feel free to apply CS2103 or CS2113 knowledge

- [x] Fix the lint warnings. I noticed there were many yellow warnings when I try building it. Do remove all the unused variables or fixed the warnings accordingly.

**Functionality**
- [x] For the date, I tested with the year 2222 (invalid year) and it worked. I think it might be good to have a React Date component for the user to select the date.

michaelseyo commented:
I've fixed the lint warnings for most of it and added comments where I could. I also found a bug in the situation where if all the other inputs are validated already but the attendees is not, the submission goes through (since attendees is not tied to react-hook-form as stated in the PR description) I've fixed the bug already. 

Separating the DataTable into components is quite difficult due to the way that TanStack React Table is made: https://tanstack.com/table/v8 and the examples provided puts everything in a file, I tried to separate but ran into quite alot of typing and props issues oops my bad for not being able to.

woowenjun99 commented:
LGTM
